### DISCUSSION TITLE
FITS Raw 0.9

### DISCUSSION THREAD

**Discussion by brampf** [brampf/fitskit#6] [Show and tell]
Since I don't have any other place to do this, please add your feedback & thoughts about FITS Raw here.

Version 0.9 is not available on TestFlight and currently provides the following features:
- Open FITS files and renders an image for the data unit of the primary HDU as image data
- Auto-detection of Grayscale, RGB and RGGB image data (particularly for the ZWO ASI294MC camera)
- Display the content of the header unit for the currently selected HDU
- Preliminary rendering of tables (Bintable only only display the first value of an array) 
- Placeholder for empty HDUs or Image HDUs with less than 1 or more than 3 dimensions
- Rendering of thumbnails for the data unit of the primary HDU of FITS files in the file browser
- Rendering of preview for the data unit of the primary HDU for apples quicklook

FITS Raw currently support only iPad OS and is tested with a rather old & slow iPad 5 on iPad OS 14.4.

In order to try out this app via Testflight, please send my your Apple ID (dev@mayworld.de) so I'll add you to the list of testers.

**Comment by todlandis** [comment]
Yes!  Sounds good!  I'm interested in using the tool for images I  pull down from online sites and others from a ZWO camera on a backyard telescope.   Look forward to trying it.

Two wishlist items ...
- Zoom and pan 
- A way to mark places on the image using the file's RA,DEC coordinate transform   (coordinates can be added to FIT files without them using astrometry.net). 
 
Here's the (so far untested) idea ...
import sims
     // check that coordinates are RA--TAN,DEC--TAN then ...
-     func pixelsToRADec( x:Double, y:Double) -> (ra:Double, dec:Double) {
        let p = simd_double2(x - CRPIX1, y - CRPIX2)
        let M = simd_double2x2([CD1_1,CD2_1], [CD1_2,CD2_2])
        let q = M * p
        return (q.x + CRVAL1, q.y + CRVAL2)
    }
    
    func RADecToPixels(ra:Double, dec:Double) ->  (x:Double, y:Double) {
        let p = simd_double2(ra - CRVAL1, dec - CRVAL2)
        let M = simd_double2x2([CD1_1,CD2_1], [CD1_2,CD2_2])
        let q = M.inverse * p
        return (q.x - CRPIX1, q.y - CRPIX2)
    }

- Tod

**Reply by brampf** [reply]
Pan & Zoom is in there.
Yesterday I pushed 0.9.1 to TestFlight which has some improvements (double-tap to fit to size) as well as allows you to select the HDU to display if there are multiple HDUs in the file.

Do you want the app to highlight parts of the image if specific header keywords are read in the file or do you want the app to write header blocks to the file based on areas you highlighted on the screen?

**Reply by brampf** [reply]
Before I dig into that coordinate thing, I will work on some auto-correction (white-balance, levels) since in astrophotography, most unedited raw files simply look plain black which is a really bad experience for thumbnails & quicklook. But in order to do so, I want to implement user based settings as any kind of "auto" will always dis-satisfy those who care the most about picture quality.

**Comment by todlandis** [comment]
I haven’t received a Testflight invite yet.  It goes to the email, yes?

Actually just what you have, with automatic gamma correction will be pretty great.   I played with FitsKit and vImagePiecewiseGamma_Planar8 ta while back.
Regarding annotations,  I’d like “draw a plus sign at (ra), (dec)”.  or “label (ra), (dec) with (text).    
- Tod

**Comment by todlandis** [comment]
I have it going now. It’s very quick, very stable  I ran it through the FITS files on  https://fits.gsfc.nasa.gov/fits_samples.html and a couple ZWO files taken inside the house without a problem.   Thunmbnails work fine.  Might be a snag with imaging a FIT sent in a message, but I'll check it again tomorrow.

A nitpick ..Pan/zoom can can drag the image off the edges of the frame.  e.g. pan left should stop when the right edge of the image is at the right edge of the bounding box.  Zoom should stop reducing when it hits 1.0.
Best,
- Tod

**Reply by brampf** [reply]
That's great to hear. I actually found a couple of issues in the advanced branch itself. As it seems I totally messed up the readout of floating point values. Furthermore I found a crash when debayering cropped images were one row of the RGGB pattern is missing.
The elastic scroll behaviour is just something which annoys me but unfortunately this seems to be an issue with SwiftUI I do not (yet!) know how to get around. Zooming beyond 100% seems to be a thing for people pixel peeping so I allowed it for now.

**Comment by brampf** [comment]
I just pushed a new build 0.9.1 to Testflight which now also renders TableHDUs and BintableHDUs instead of the placeholders.

**Comment by brampf** [comment]
I just pushed new build 0.9.1 to Testflight which finally no longer bounces the images when scrolling

**Comment by mburge** [comment]
Are you going to release an App Store version?  Can I join the TestFlight?
